### DISCUSSION TITLE
Creating new tracks & changing instrument/sound programatically

### DISCUSSION THREAD

**Discussion by zopieux** [musescore/MuseScore#19431] [Q&A] [answered]
Hi,

I'm trying to programmatically create new parts (this works) and change their instrument (not working) directly in [this C++ function](https://github.com/musescore/MuseScore/blob/master/src/converter/internal/convertercontroller.cpp#L67-L100).

```cpp
    // Added in .h
    INJECT(notation::IInstrumentsRepository, instrumentsRepository)

    // Added in .c ConverterController::fileConvert
    mu::notation::IMasterNotationPtr master = notationProject->masterNotation();
    mu::notation::INotationPtr notation = notationProject->masterNotation()->notation();
    mu::engraving::MasterScore *score = notationProject->masterNotation()->masterScore();
    mu::notation::INotationPartsPtr notationParts = notation->parts();

    PartInstrumentList partList;
    IDList originalIDs;
    size_t idx = 0;
    for (const auto *p : notationParts->partList())
    {
        originalIDs.emplace_back(p->id());
        partList.append(PartInstrument{
            .partId = p->id(),
            .instrumentTemplate = instrumentsRepository()->instrumentTemplate(p->instrumentId().toStdString()),
            .isExistingPart = true,  // exisiting
            .isSoloist = p->soloist(),
        });
        partList.append(PartInstrument{
            .partId = notationParts->partList().size() + (idx++),
            .instrumentTemplate = instrumentsRepository()->instrumentTemplate(p->instrumentId().toStdString()),
            .isExistingPart = false,  // new
            .isSoloist = p->soloist(),
        });
    }
    master->parts()->setParts(partList, master->parts()->scoreOrder());
    score->doLayout();
```

While this correctly creates new parts for each existing part, it seems like the `instrumentRepository` is not working as intended. The instrument template is the `dummy` empty one, and when doing playback, there is no instrument found.

1. Is this the correct approach to create a new empty part for each existing expart?
2. How can I fix the above code to assign the same instrument & SoundFont to the newly created parts? It looks like the `instrumentRepository` is not correctly initialized because I'm not using it appropriately. But I don't know how to otherwise fill in the `instrumentTemplate` field.

Thanks.

**Comment by sammik** [comment]
To change instrument, there is `part.setInstrument(instrument)` https://github.com/musescore/MuseScore/blob/eff81f3f32558a8634ea279d8e0632594487df59/src/engraving/dom/part.h#L143 isnt it?

**Reply by zopieux** [reply]
Thanks, most likely, yes but it seems like the instrument repository is not initialized so it's empty. I can't wrap my head around the dependency injection stuff. Can't I just instantiate one manually?

**Reply by cbjeukendrup** [reply]
The instance of `IInstrumentsRepository` is instantiated here: https://github.com/musescore/MuseScore/blob/a13d8b4dbf3c52267396d92aae27d98b6666a5f4/src/notation/notationmodule.cpp#L106-L110
It is initialised here: https://github.com/musescore/MuseScore/blob/a13d8b4dbf3c52267396d92aae27d98b6666a5f4/src/notation/notationmodule.cpp#L229
This is the instance that is retrieved using dependency injection.

**Reply by cbjeukendrup** [reply, marked as answer]
If the repository is indeed still empty, you're most likely running a non-installed build: you're using the executable from the build folder instead of the one from the install folder specified using `CMAKE_INSTALL_PREFIX`.

**Reply by zopieux** [reply]
@cbjeukendrup Thanks for bearing with me. Indeed I'm running the binary straight out of `ninja` since I'm iterating on my code patch. I'll try to "install" properly or hack around search paths.

As for dependency injection, what's unclear to me is whether the `convertcontroller` context in which I'm running this code has access to the correctly initialized instance, or if more boilerplate code is necessary. My guess is that module initialization (including `notationmodule`) is done early in the main binary and so I don't need to take extra precautions. It's just not working because it can't find the instruments (wrong run location per your previous comment).

**Reply by cbjeukendrup** [reply]
Yes, those module initialisation methods are called in `App::run` (app.cpp), which is basically the only interesting content of the `main` function. So as long as you're just working on the MuseScore app, initialisation should not be a problem. Only if you're writing a new unit tests suite or creating your own app using some code from MuseScore, you need to be careful (and also when working in `onInit` of some other module, of course).

**Reply by zopieux** [reply]
Thanks, it was indeed an issue with the repository not finding the instrument files in its configured prefix.
